examples/example_mnist_prune.py

```diff
@@ -0,0 +1,227 @@
+    x = QConv2D(
+        32, (2, 2), strides=(2,2),
+        kernel_quantizer=quantized_bits(4,0,1),
+        bias_quantizer=quantized_bits(4,0,1),
+        name="conv2d_0_m")(x)
+    x = QActivation("quantized_relu(4,0)", name="act0_m")(x)
+    x = QConv2D(
+        64, (3, 3), strides=(2,2),
+        kernel_quantizer=quantized_bits(4,0,1),
+        bias_quantizer=quantized_bits(4,0,1),
+        name="conv2d_1_m")(x)
+    x = QActivation("quantized_relu(4,0)", name="act1_m")(x)
+    x = QConv2D(
+        64, (2, 2), strides=(2,2),
+        kernel_quantizer=quantized_bits(4,0,1),
+        bias_quantizer=quantized_bits(4,0,1),
+        name="conv2d_2_m")(x)
+    x = QActivation("quantized_relu(4,0)", name="act2_m")(x)
+    x = Flatten()(x)
+    x = QDense(num_classes, kernel_quantizer=quantized_bits(4,0,1),
+               bias_quantizer=quantized_bits(4,0,1),
+               name="dense")(x)
+    x = Activation("softmax", name="softmax")(x)
+
+    model = Model(inputs=[x_in], outputs=[x])
+    return model
+
+
+def build_layerwise_model(input_shape, **pruning_params):
+    return Sequential([
+        prune.prune_low_magnitude(
+            QConv2D(
+                32, (2, 2), strides=(2,2),
+                kernel_quantizer=quantized_bits(4,0,1),
+                bias_quantizer=quantized_bits(4,0,1),
+                name="conv2d_0_m"),
+            input_shape=input_shape,
+            **pruning_params),
+        QActivation("quantized_relu(4,0)", name="act0_m"),
+        prune.prune_low_magnitude(
+            QConv2D(
+                64, (3, 3), strides=(2,2),
+                kernel_quantizer=quantized_bits(4,0,1),
+                bias_quantizer=quantized_bits(4,0,1),
+                name="conv2d_1_m"),
+            **pruning_params),
+        QActivation("quantized_relu(4,0)", name="act1_m"),
+        prune.prune_low_magnitude(
+            QConv2D(
+                64, (2, 2), strides=(2,2),
+                kernel_quantizer=quantized_bits(4,0,1),
+                bias_quantizer=quantized_bits(4,0,1),
+                name="conv2d_2_m"),
+            **pruning_params),
+        QActivation("quantized_relu(4,0)", name="act2_m"),
+        Flatten(),
+        prune.prune_low_magnitude(
+            QDense(
+                num_classes, kernel_quantizer=quantized_bits(4,0,1),
+                bias_quantizer=quantized_bits(4,0,1),
+                name="dense"),
+            **pruning_params),
+        Activation("softmax", name="softmax")
+  ])
+
+
+def print_model_sparsity(pruned_model):
+    """Prints sparsity for the pruned layers in the model."""
+
+    def _get_sparsity(weights):
+        return 1.0 - np.count_nonzero(weights) / float(weights.size)
+
+    print("Model Sparsity Summary ({})".format(pruned_model.name))
+    print("--")
+    for layer in pruned_model.layers:
+        if isinstance(layer, pruning_wrapper.PruneLowMagnitude):
+            prunable_weights = layer.layer.get_prunable_weights()
```

Comment:
I think it is good to put it in utils.py

Also, make it supports ```weights = layer.layer.get_weights()```

What do you think?

Reply:
If we check all weights (with `layer.get_weights()`) that will give us the total sparsity, which may be misleading since not all layers with weights can be pruned. We can extend the current check to include the layers which are prunable, but the user chose not to prune them (i.e., they are in `PruneRegistry`). Would that be ok?

As for supporting clusterable layers, `QConv2D(QConv2D, ClusterableLayer, PruableLayer)` is the only way I see for now. I was hoping for a mechanism for registering layers with (Prune|Clustering)Registry, but there is none currently. Also, clustering is not part of any released version of tfmot, nor is currently part of public API in the master on github (see [this](https://github.com/tensorflow/model-optimization/commit/f1f5afb05e2c41970e0bee1624f08503cd71c0c8) commit) which makes me think we should wait a little bit before supporting it.

Reply:
> If we check all weights (with layer.get_weights()) that will give us the total sparsity, which may be misleading since not all layers with weights can be pruned. We can extend the current check to include the layers which are prunable, but the user chose not to prune them (i.e., they are in PruneRegistry). Would that be ok?

Yes. thanks!

> As for supporting clusterable layers, QConv2D(QConv2D, ClusterableLayer, PruableLayer) is the only way I see for now. I was hoping for a mechanism for registering layers with (Prune|Clustering)Registry, but there is none currently. Also, clustering is not part of any released version of tfmot, nor is currently part of public API in the master on github (see this commit) which makes me think we should wait a little bit before supporting it.

answered in another thread. please check.